### ISSUE TITLE
Dynamo unsupported: Dynamic slicing on data-dependent value is not supported

### ISSUE THREAD

**Issue by BowenBao** [pytorch/pytorch#123592]
### 🐛 Describe the bug

A simplified repro is as follows. 

```python
import torch


def func(x, img_size):
    x = x[:img_size]
    return x

x = torch.randn(10, 3, 224, 224)
img_size = torch.tensor(5)
torch._dynamo.export(func)(x, img_size)
```

```
  File "/pytorch/torch/_dynamo/symbolic_convert.py", line 1391, in BUILD_SLICE
    self.push(SliceVariable(items))
  File "/pytorch/torch/_dynamo/variables/lists.py", line 597, in __init__
    unimplemented("Dynamic slicing on data-dependent value is not supported")
  File "/pytorch/torch/_dynamo/exc.py", line 204, in unimplemented
    raise Unsupported(msg)
torch._dynamo.exc.Unsupported: Dynamic slicing on data-dependent value is not supported

from user code:
   File "/repro_datadependent_slice.py", line 13, in func
    x = x[:img_size]

Set TORCH_LOGS="+dynamo" and TORCHDYNAMO_VERBOSE=1 for more information
```

### Versions

torch                     2.4.0a0+git4946558

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim @chauhang @penguinwu @bobrenjc93 @bdhirsh @anijain2305 @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @avikchaudhuri @gmagogsfm @zhxchen17 @tugsbayasgalan @angelayi @suo @ydwu4

**Comment by thiagocrepaldi** [comment]
https://github.com/pytorch/pytorch/pull/91341 was similar to this, but somehow works

**Comment by angelayi** [comment]
I think this is similar to if we were to slice on an unbacked symint, like
```
def func(x, img_size):
    idx = img_size.item()
    x = x[:idx]
    return x
```

In which case you would need to add some `torch._check`s
```
def func(x, img_size):
    idx = img_size.item()
    torch._check_is_size(idx)
    torch._check(idx < x.size(0))
    x = x[:idx]
    return x
```

**Comment by thiagocrepaldi** [comment]
> ```python
> def func(x, img_size):
>     idx = img_size.item()
>     torch._check_is_size(idx)
>     torch._check(idx < x.size(0))
>     x = x[:idx]
>     return x
> ```

The code above still fails, although differently:

```python
def func(x, img_size):
    idx = img_size.item()
    torch._check_is_size(idx)
    torch._check(idx < x.size(0))
    x = x[:idx]
    return x

x = torch.randn(10, 3, 224, 224)
img_size = torch.tensor(5)
torch._dynamo.export(func)(x, img_size)
```

Error

```bash
Traceback (most recent call last):
  File "/opt/pytorch/torch/fx/graph_module.py", line 304, in __call__
    return super(self.cls, obj).__call__(*args, **kwargs)  # type: ignore[misc]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/pytorch/torch/nn/modules/module.py", line 1532, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/pytorch/torch/nn/modules/module.py", line 1541, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<eval_with_key>.1", line 13, in forward
    mul_1 = -1 * size_1
                 ^^^^^^
UnboundLocalError: cannot access local variable 'size_1' where it is not associated with a value

Call using an FX-traced Module, line 13 of the traced Module's generated forward function:
    _assert_scalar_default = torch.ops.aten._assert_scalar.default(le, 'Deferred runtime assertion failed -u0 <= 0');  le = None
    mul_1 = -1 * size_1

~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
    add = item + mul_1;  mul_1 = None

    lt_1 = add < 0;  add = None

Traceback (most recent call last):
  File "/opt/pytorch/repro_dyn_slice.py", line 17, in <module>
    torch._dynamo.export(func)(x, img_size)
  File "/opt/pytorch/torch/_dynamo/eval_frame.py", line 1237, in inner
    result_traced = opt_f(*args, **kwargs)
                    ^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/pytorch/torch/_dynamo/eval_frame.py", line 410, in _fn
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
  File "/opt/pytorch/repro_dyn_slice.py", line 8, in func
    def func(x, img_size):
  File "/opt/pytorch/torch/_dynamo/eval_frame.py", line 410, in _fn
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
  File "/opt/pytorch/torch/_dynamo/external_utils.py", line 36, in inner
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
  File "/opt/pytorch/torch/_dynamo/eval_frame.py", line 1203, in result_capturing_wrapper
    graph_captured_result = torch.func.functional_call(
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/pytorch/torch/_functorch/functional_call.py", line 143, in functional_call
    return nn.utils.stateless._functional_call(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/pytorch/torch/nn/utils/stateless.py", line 263, in _functional_call
    return module(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/pytorch/torch/nn/modules/module.py", line 1532, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/pytorch/torch/nn/modules/module.py", line 1541, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/pytorch/torch/fx/_lazy_graph_module.py", line 123, in _lazy_forward
    return self(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/opt/pytorch/torch/fx/graph_module.py", line 737, in call_wrapped
    return self._wrapped_call(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/pytorch/torch/fx/graph_module.py", line 315, in __call__
    raise e.with_traceback(None)  # noqa: TRY200
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
UnboundLocalError: cannot access local variable 'size_1' where it is not associated with a value
(ptca) root@a6e6443516b9:/opt/pytorch# 
```

but using  torch.export.export` exports, at least

```python
class MySlice(torch.nn.Module):
    def forward(self, x, img_size):
        idx = img_size.item()
        torch._check_is_size(idx)
        torch._check(idx < x.size(0))
        x = x[:idx]
        return x

x = torch.randn(10, 3, 224, 224)
img_size = torch.tensor(5)
torch.export.export(MySlice(), args=(x, img_size))
```

Any advice for the `torch._dynamo.export` path?

**Comment by ezyang** [comment]
This seems just like a bug

**Comment by bhack** [comment]
I have another standalone case with with the same error at https://github.com/pytorch/pytorch/issues/4073#issuecomment-2163980614

It work with compile but when using `fullgraph=True` I got the same error.

**Comment by bhack** [comment]
Also probably this is a duplicate of https://github.com/pytorch/pytorch/issues/123592

**Comment by raziel** [comment]
Hi @shazqadeer , @angelayi  any update?
This makes it hard to work with "stateful" models, like KVCaches, where we need to slice and dice tensors for efficient updates.

**Comment by avikchaudhuri** [comment]
@raziel here's a workaround, does that work?

```
import torch

    class M(torch.nn.Module):
        def forward(self, x, img_size):
            # x = x[:img_size]
            x = x.narrow(0, 0, img_size.item())
            return x

    x = torch.randn(10, 3, 4)
    img_size = torch.tensor(5)
    m = M()
    inp = (x, img_size)
    ep = torch.export.export(m, inp)
    print(m(*inp))
    print(ep.module()(*inp))
    print(x[:img_size])
```

The `narrow` avoids data-dependent shape errors without needing additional hints from the user.

Aside: I'm curious why you're using `_dynamo.export()`, as it is not a public API.

**Comment by raziel** [comment]
Thanks @avikchaudhuri  we'll try it out.

We're not using ```dynamic_export```, we're using the public ```torch.export```

Taking the example above with torch.export:

```python
class MySlice(torch.nn.Module):
    def forward(self, x, img_size):
        # idx = img_size.item()
        # torch._check_is_size(idx)
        # torch._check(idx < x.size(0))
        x = x[:img_size]
        return x

x = torch.randn(10, 3, 224, 224)
img_size = torch.tensor(5)
torch.export.export(MySlice(), args=(x, img_size))
```

This results in eventually calling torch._dynamo.export which fails as described in the issue:

```python
Unsupported: Dynamic slicing on data-dependent value is not supported

from user code:
   File "<ipython-input-5-86521f595616>", line 8, in forward
    x = x[:img_size]
```

When I replace your workaround as:

```python
class MySlice(torch.nn.Module):
    def forward(self, x, img_size):
        # idx = img_size.item()
        # torch._check_is_size(idx)
        # torch._check(idx < x.size(0))
        # x = x[:img_size]
        x = x.narrow(0, 0, img_size.item())
        return x

x = torch.randn(10, 3, 224, 224)
img_size = torch.tensor(5)
torch.export.export(MySlice(), args=(x, img_size))
```

You get the following error:

```python
GuardOnDataDependentSymNode: Could not guard on data-dependent expression Eq(u0, 0) (unhinted: Eq(u0, 0)).  (Size-like symbols: none)

Potential framework code culprit (scroll up for full backtrace):
  File "/usr/local/lib/python3.10/dist-packages/torch/_subclasses/functional_tensor.py", line 102, in __new__
    out = torch.Tensor._make_wrapper_subclass(  # type: ignore[arg-type, attr-defined]
...
While executing %x : [num_users=1] = call_method[target=narrow](args = (%l_x_, 0, 0, %item), kwargs = {})
Original traceback:
  File "<ipython-input-6-8d1a07f234bb>", line 9, in forward
    x = x.narrow(0, 0, img_size.item())
```

The same error occurs running your copy-pasted code.

**Comment by avikchaudhuri** [comment]
Strange, it doesn't error for me. I'm working off trunk though. I believe @ezyang landed some changes a while back that don't result in that error any more. Maybe your PT2 version is old? If so, maybe uncomment the lines
```
        # idx = img_size.item()
        # torch._check_is_size(idx)
        # torch._check(idx < x.size(0))
```
and try again (using `idx` instead of `img_size.item()` in the `narrow` call)?

**Comment by avikchaudhuri** [comment]
I also recently added some suggested fixes in the error message for `GuardOnDataDependentSymNode` errors which basically suggest these `torch._check` calls, but unfortunately it only works with `torch.export.export(..., strict=False)`, which doesn't go through Dynamo, right now. (This is FYI, not relevant if you don't use that flag right now, but I plan to do this for `strict=True` as well at some point. More details here: https://github.com/pytorch/pytorch/pull/125378)

**Comment by raziel** [comment]
I'm woking out of 2.4.0+cu121 --also tried 2.3.1.
I'm assuming if you can try 2.4 version you should see it.

So maybe something @ezyang fixed in trunk? Do you know when that fix would be released?

**Comment by raziel** [comment]
Adding the checks seem to complete the workaround (though it'd be great if they're needed after @ezyang 's fixed you mentioned):

```python
import torch

class MySlice(torch.nn.Module):
    def forward(self, x, img_size):
        idx = img_size.item()
        torch._check_is_size(idx)
        torch._check(idx < x.size(0))
        x = x.narrow(0, 0, idx)
        return x

x = torch.randn(10, 3, 224, 224)
img_size = torch.tensor(5)
print(torch.export.export(MySlice(), args=(x, img_size)))
```

Exports:

```python
ExportedProgram:
    class GraphModule(torch.nn.Module):
        def forward(self, x: "f32[10, 3, 224, 224]", img_size: "i64[]"):
            # File: <ipython-input-9-f5bd7b99bce1>:5 in forward, code: idx = img_size.item()
            item: "Sym(u0)" = torch.ops.aten.item.default(img_size);  img_size = None
            
            # 
            ge: "Sym(u0 >= 0)" = item >= 0
            scalar_tensor_default: "f32[]" = torch.ops.aten.scalar_tensor.default(ge);  ge = None
            _assert_async_msg = torch.ops.aten._assert_async.msg(scalar_tensor_default, 'item is outside of inline constraint [0, 9].');  scalar_tensor_default = None
            le_1: "Sym(u0 <= 9)" = item <= 9
            scalar_tensor_default_1: "f32[]" = torch.ops.aten.scalar_tensor.default(le_1);  le_1 = None
            _assert_async_msg_1 = torch.ops.aten._assert_async.msg(scalar_tensor_default_1, 'item is outside of inline constraint [0, 9].');  scalar_tensor_default_1 = None
            
            # 
            sym_constrain_range_for_size_default = torch.ops.aten.sym_constrain_range_for_size.default(item)
            
            # File: /usr/local/lib/python3.10/dist-packages/torch/_export/passes/add_runtime_assertions_for_constraints_pass.py:157 in call, code: node.meta["stack_trace"] = "".join(traceback.format_stack(limit=1))
            sym_constrain_range = torch.ops.aten.sym_constrain_range.default(item, min = 0, max = 9)
            mul: "Sym(-u0)" = -1 * item
            le: "Sym(-u0 <= 0)" = mul <= 0;  mul = None
            _assert_scalar = torch.ops.aten._assert_scalar.default(le, "Runtime assertion failed for expression -u0 <= 0 on node 'le'");  le = None
            lt: "Sym(u0 < 10)" = item < 10
            _assert_scalar_1 = torch.ops.aten._assert_scalar.default(lt, "Runtime assertion failed for expression u0 < 10 on node 'lt_1'");  lt = None
            
            # File: <ipython-input-9-f5bd7b99bce1>:8 in forward, code: x = x.narrow(0, 0, idx)
            add: "Sym(u0)" = 0 + item;  item = None
            slice_1: "f32[u0, 3, 224, 224]" = torch.ops.aten.slice.Tensor(x, 0, 0, add);  x = add = None
            return (slice_1,)
            
Graph signature: ExportGraphSignature(input_specs=[InputSpec(kind=<InputKind.USER_INPUT: 1>, arg=TensorArgument(name='x'), target=None, persistent=None), InputSpec(kind=<InputKind.USER_INPUT: 1>, arg=TensorArgument(name='img_size'), target=None, persistent=None)], output_specs=[OutputSpec(kind=<OutputKind.USER_OUTPUT: 1>, arg=TensorArgument(name='slice_1'), target=None)])
Range constraints: {u0: VR[0, 9], u1: VR[0, 9], u2: VR[0, 9]}
```

**Comment by gilfree** [comment]
The narrow workaround fails to work, if slice is not on the first dimension, which is common in cases of embedding in transformer decoder style networks, when running them batched:

```python

import torch
class MySlice(torch.nn.Module):
    def forward(self, x, seq_len):
        x = x.narrow(1, 0, seq_len.item())
        return x

x = torch.randn(10, 7)
seq_len = torch.tensor(5)
torch.export.export(MySlice(), args=(x, seq_len))
```

Note the difference from above is just that I narrow on dim 1 and not 0 (This happens on a recent nightly also), it fails with:

```
torch._dynamo.exc.UserError: Could not guard on data-dependent expression Eq(7, u0) (unhinted: Eq(7, u0)).  (Size-like symbols: u0)
```

As far as I can tell the source of the error is various is_contiguous checks that fail (I can get a traces to guard_size_oblivious with file '/pytorch/c10/core/Contiguity.h'), when I put breakpoint in `raise self._make_data_dependent_error` in the `symbolic_shapes.py` file.)
 
It would have been great if there was no need to the narrow workaround, which requires to change a very common code pattern, but in even narrow fails to work.

Note that there is a costly workaround:

```
def my_slice(x,dim,end):
    x = x.transpose(0, dim).contiguous()
    y= x.narrow(0, 0, end)
    return y.transpose(0, dim)
```    

But it's a "magical" solution, and I have no idea if it will hold in the next nightly.

Is there anything that can be done to support standard slicing? Or is it a hard to overcome limitation? If there is something to be done, if someone is willing to guide me how to approch this, I'll be happy to try and help.

**Comment by gilfree** [comment]
> @gilfree This should work:
> 
> ```python
>         class MySlice(torch.nn.Module):
>             def forward(self, x, seq_len):
>                 y = seq_len.item()  # refactor into local variable
>                 torch._check(x.shape[1] != y)  # add missing assumption
>                 x = x.narrow(1, 0, y)
>                 return x
> 
>         x = torch.randn(10, 7)
>         seq_len = torch.tensor(5)
>         torch.export.export(MySlice(), args=(x, seq_len))
> ```
> 
> With `strict=False`, you would get a suggestion for a fix after you separate `seq_len.item()` into a local variable `y`, asking to introduce either `torch._check(7 == y)` or `torch._check(7 != y)`. The `7` is just `x.shape[1]` here.
> 
> With our WIP `torch.export._draft_export.draft_export` you'd get a graph with the check added automatically, which you can use to unblock; although in general this is unsafe to use because it automatically specializes on conditions satisfied by data-dependent values.

Thanks, but that is an incorrect assumption. In transformers, we have the sequence len increasing until it reaches the full sequence, so the graph should be correct for y<=x.shape[1] and not only y<x.shape[1]

**Comment by avikchaudhuri** [comment]
@ezyang any ideas?

**Comment by ezyang** [comment]
This feels similar to https://github.com/pytorch/pytorch/issues/120288

**Comment by ezyang** [comment]
So I ran the repro and it turns out there are two independent problems, but at least in the minimal repro, it is failing for a dumb reason (however, I'm not sure if it will work in the full context.)

First, with this patch:

```
diff --git a/torch/fx/passes/shape_prop.py b/torch/fx/passes/shape_prop.py
index 341bfd5e409..3e10dcf43ae 100644
--- a/torch/fx/passes/shape_prop.py
+++ b/torch/fx/passes/shape_prop.py
@@ -45,7 +45,7 @@ def _extract_tensor_metadata(
 
     memory_format = None
 
-    if include_contiguity and not is_sparse_any(result):
+    if False and not is_sparse_any(result):
         memory_formats = {
             torch.contiguous_format,
             torch.channels_last,
```

then this works:

```
TORCH_FAKE_TENSOR_DISPATCH_CACHE=0 python wu.py
```

The only reason the guard was occurring was because we were stupidly trying to query contiguity on every intermediate result, which doesn't work here because we don't know (cc @eellison, this was your intuition right? So we just don't query contiguity and it's fine.)

Now, I have to disable the fake tensor cache, because the cache itself accidentally triggers are guard it shouldn't (cc @aorenste )

```
  File "python_variable_methods.cpp", line 0, in torch::autograd::THPVariable_set_(_object*, _object*, _object*)
  File "??", line 0, in at::_ops::set__source_Storage_storage_offset::call(at::Tensor&, c10::Storage, c10::SymInt, c10::ArrayRef<c10::
SymInt>, c10::ArrayRef<c10::SymInt>)                                                                                                  
  File "offloadstuff.c", line 0, in c10::impl::BoxedKernelWrapper<at::Tensor& (at::Tensor&, c10::Storage, c10::SymInt, c10::ArrayRef<c
10::SymInt>, c10::ArrayRef<c10::SymInt>), void>::call(c10::BoxedKernel const&, c10::OperatorHandle const&, c10::DispatchKeySet, at::Te
nsor&, c10::Storage, c10::SymInt, c10::ArrayRef<c10::SymInt>, c10::ArrayRef<c10::SymInt>)
  File "PyInterpreter.cpp", line 0, in torch::detail::(anonymous namespace)::ConcretePyInterpreterVTable::python_dispatcher(c10::Opera
torHandle const&, c10::DispatchKeySet, std::vector<c10::IValue, std::allocator<c10::IValue> >*) const
  File "offloadstuff.c", line 0, in c10::OperatorHandle::callBoxedForDispatchKey(c10::DispatchKey, std::vector<c10::IValue, std::alloc
ator<c10::IValue> >&) const                                                                                                           
  File "RegisterMeta.cpp", line 0, in c10::impl::make_boxed_from_unboxed_functor<c10::impl::detail::WrapFunctionIntoFunctor_<c10::CompileTimeFunctionPointer<at::Tensor& (at::Tensor&, c10::Storage, c10::SymInt, c10::ArrayRef<c10::SymInt>, c10::ArrayRef<c10::SymInt>), &at::(anonymous namespace)::(anonymous namespace)::wrapper_Meta_source_Storage_storage_offset_set_>, at::Tensor&, c10::guts::typelist::typelist<at::Tensor&, c10::Storage, c10::SymInt, c10::ArrayRef<c10::SymInt>, c10::ArrayRef<c10::SymInt> > >, false>::call(c10::OperatorKernel*, c10::OperatorHandle const&, c10::DispatchKeySet, std::vector<c10::IValue, std::allocator<c10::IValue> >*)           
  File "RegisterMeta.cpp", line 0, in at::(anonymous namespace)::(anonymous namespace)::wrapper_Meta_source_Storage_storage_offset_set
_(at::Tensor&, c10::Storage, c10::SymInt, c10::ArrayRef<c10::SymInt>, c10::ArrayRef<c10::SymInt>)
  File "??", line 0, in at::native::set_storage_meta__symint(at::Tensor&, c10::Storage, c10::SymInt, c10::ArrayRef<c10::SymInt>, c10::
ArrayRef<c10::SymInt>)                                                                                                                
  File "??", line 0, in c10::TensorImpl::is_contiguous_custom(c10::MemoryFormat) const
  File "??", line 0, in c10::SymbolicShapeMeta::init_is_contiguous() const
  File "??", line 0, in c10::SymbolicShapeMeta::compute_contiguous() const
  File "??", line 0, in c10::SymBool::guard_size_oblivious(char const*, long) const
  File "offloadstuff.c", line 0, in torch::impl::PythonSymNodeImpl::guard_size_oblivious(char const*, long)
```

It looks like `set_` queries contiguity, which also triggers the problem like before.

So why am I unsure if it will work in the full context? The main problem is that if you end up using the result of narrow in a context where we need to probe its contiguity, it will explode again, even if we've removed the two mentioned contiguity tests above.

**Comment by ezyang** [comment]
https://github.com/pytorch/pytorch/pull/144471 as a possible way to fix this

**Comment by gilfree** [comment]
@ezyang, maybe there something I can do to assert/always reach the pessimistic case of non contiguous? 

As my target is export I can later fix pessimistic patterns, but export failing gives me no graph to work with.

**Comment by ezyang** [comment]
I landed https://github.com/pytorch/pytorch/pull/144471 . Get a nightly, and then do `torch._check_is_size(y, max=x.shape[1])` and it should work now

**Comment by angelayi** [comment]
Closing as there seems to be workarounds/fixes for the issues listed above:

For this type of code:
```python
class M(torch.nn.Module):
    def forward(self, x: torch.Tensor, img_size: torch.Tensor):
        x = x[:img_size]
        return x
```
There are two workarounds:
```python
class M(torch.nn.Module):
    def forward(self, x, img_size):
        return x.narrow(0, 0, img_size.item())

class M(torch.nn.Module):
    def forward(self, x, img_size):
        idx = img_size.item()
        torch._check_is_size(idx)
        torch._check(idx < x.size(0))
        x = x[:idx]
        return x
```

In the case where we are narrow-ing on a different dimension:
```python
class M(torch.nn.Module):
    def forward(self, x, img_size):
        y = img_size.item()
        torch._check_is_size(y, max=x.shape[1])
        return x.narrow(1, 0, y)
```

**Comment by laithsakka** [comment]
@pianpwk PR fixed this good
